File: LeetCode/Java/src/solutions/p1402_ReducingDishes/P1402.java
package solutions.p1402_ReducingDishes;

import java.util.Arrays;

public class P1402 {
    public int maxSatisfaction(int[] satisfaction) {
        /*
         * Solution v1: Greedy
         *
         * Runtime: 1 ms @ (beats) 100%
         * Memory Usage: 40.3 MB @ (beats) 91.75%
         */

        Arrays.sort(satisfaction);

        int ans = 0;
        int curr_sati = 0;

        for (int i = satisfaction.length - 1; i >= 0; i--) {
            curr_sati += satisfaction[i];
            if (curr_sati < 0) {
                break;
            }
            ans += curr_sati;
        }
        return ans;
    }
}

File: LeetCode/Python/solutions/p1402.py
from typing import List


class Solution:
    def maxSatisfaction(self, satisfaction: List[int]) -> int:
        #
        # Solution v1: Greedy
        #
        # Runtime: 43 ms @ (beats) 69.30%
        # Memory Usage: 13.9 MB @ (beats) 46.84%
        #
        satisfaction.sort(reverse=True)
        ans = 0
        curr_sati = 0
        for dish in satisfaction:
            curr_sati += dish
            if curr_sati < 0:
                break
            # this step is the core
            # accumulate addition will make sure
            # every dish be counted in correct times.
            ans += curr_sati

        return ans


if __name__ == "__main__":
    solution = Solution()

    satisfaction = [-1, -8, 0, 5, -9]
    print(f"return: {solution.maxSatisfaction(satisfaction)}")  # 14

    satisfaction = [4, 3, 2]
    print(f"return: {solution.maxSatisfaction(satisfaction)}")  # 20

    satisfaction = [-1, -4, -5]
    print(f"return: {solution.maxSatisfaction(satisfaction)}")  # 0
